data_structure/trie_modern.cpp

```diff
@@ -0,0 +1,156 @@
+#include <iostream>
+#include <memory>  // for std::shared_ptr<>
+#include <string>  // for std::string class
```

Comment:
the `<string>` header is specific to Microsoft Visual C++. To make this generic, please use the following:
```c
#ifdef _MSC_VER
#include <string>
#else
#include <cstring>
#endif
```